.buildkite/scripts/pipelines/trigger_version_dependent_jobs/pipeline.ts

```diff
@@ -0,0 +1,198 @@
+ * Should be triggered for all open branches in the versions.json: 7.x, 8.x
+ */
+function getArtifactSnapshotPipelineTriggers() {
+  // Trigger for all named branches
+  const versions = getVersionsFile();
+  const branches = versions.map((v) => v.branch);
+
+  return branches.map((branch) => {
+    return {
+      trigger: pipelineSets['artifacts-snapshot'],
+      async: true,
+      label: `Triggering snapshot artifact builds for ${branch}`,
+      build: {
+        message: process.env.MESSAGE || `Snapshot artifact build for ${branch}`,
+        branch,
+        commit: 'HEAD',
+        env: {
+          DRY_RUN: process.env.DRY_RUN,
+        },
+      },
+    } as BuildkiteTriggerStep;
+  });
+}
+
+/**
+ * This pipeline creates Kibana artifacts for branches that are not the current main.
+ * Should be triggered for all open branches in the versions.json: 7.x, 8.x, but not main.
+ */
+function getArtifactStagingPipelineTriggers() {
+  // Trigger for all branches, that are not current minor+major
+  const versions = getVersionsFile();
+  const branches = versions.filter((v) => !v.currentMajor && !v.currentMinor).map((v) => v.branch);
+
+  return branches.map((branch) => {
+    return {
+      trigger: pipelineSets['artifacts-staging'],
+      async: true,
+      label: `Triggering staging artifact builds for ${branch}`,
+      build: {
+        message: process.env.MESSAGE || `Staging artifact build for ${branch}`,
+        branch,
+        commit: 'HEAD',
+        env: {
+          DRY_RUN: process.env.DRY_RUN,
+        },
+      },
+    } as BuildkiteTriggerStep;
+  });
+}
+
+/**
+ * This pipeline checks if there are any changes in the incorporated $BEATS_MANIFEST_LATEST_URL (beats version)
+ * and triggers a staging artifact build.
+ * Should be triggered only for the active minor versions that are not `main` and not `7.17`.
+ *
+ * TODO: we could basically do the check logic of .buildkite/scripts/steps/artifacts/trigger.sh in here, and remove kibana-artifacts-trigger
+ */
+function getArtifactBuildTriggers() {
+  const branches = getVersionsFile()
+    .filter((v) => v.branch !== 'main' && v.branch !== '7.17')
+    .map((v) => v.branch);
+
+  return branches.map(
+    (branch) =>
+      ({
+        trigger: pipelineSets['artifacts-trigger'],
+        async: true,
+        label: `Triggering artifact build for ${branch}`,
+        build: {
+          message: process.env.MESSAGE || `Artifact build for ${branch}`,
+          branch,
+          commit: 'HEAD',
+          env: {
+            DRY_RUN: process.env.DRY_RUN,
+          },
+        },
+      } as BuildkiteTriggerStep)
+  );
+}
+
+function getVersionsFile(): Array<{
+  version: string;
+  branch: string;
+  previousMajor?: boolean;
+  previousMinor?: boolean;
+  currentMajor?: boolean;
+  currentMinor?: boolean;
+}> {
+  try {
+    const versions = JSON.parse(
+      fs.readFileSync(path.join(getKibanaDir(), 'versions.json')).toString()
+    );
+    return versions.versions;
```

Comment:
It could be good to adjust the data structure here, so that we can avoid the `filter` and `map` in calling functions. Easier to maintain and extend for more pipelines in the future. Something like this comes to mind for me, idk if it's the "best" though

```js
const versions = {
  branches: ['main', '7.17', ...],
  previousMajor: '7.17',
  '7.17': {
    "version": "7.17.23"
    "branch": "7.17"
  },
  ...
}

const prevMajor = versions[versions.previousMajor];
```

Reply:
I'll see if I can adopt something like this, thanks for the suggestion

Reply:
@Ikuni17 - Updated to use a nicer versions model, what do you think?

Reply:
@delanni lgtm, thanks for updating it!